### ISSUE TITLE
h2spec 5.1.2 failed when reverse proxy

### ISSUE THREAD

**Issue by zlm2012** [h2o/h2o#341]
I ran [h2spec](https://github.com/summerwind/h2spec) against a reverse-proxied host on h2o and got this failure:

```
===============================================================================
Failed tests
===============================================================================

  5.1. Stream States
    5.1.2. Stream Concurrency
      × Sends HEADERS frames that causes their advertised concurrent stream limit to be exceeded
        - The endpoint MUST treat this as a stream error (Section 5.4.2) of type PROTOCOL_ERROR or REFUSED_STREAM
          Expected: GOAWAY frame (ErrorCode: PROTOCOL_ERROR)
                    RST_STREAM frame (ErrorCode: PROTOCOL_ERROR)
                    GOAWAY frame (ErrorCode: REFUSED_STREAM)
                    RST_STREAM frame (ErrorCode: REFUSED_STREAM)
                    Connection close
            Actual: HEADERS frame
```
(And if the backend added `Link` headers to support server-push, the `Actual` field above would be `PUSH_PROMISE frame`. Standard file serving is just okay to satisfy requirements of h2spec.)

IMHO this should be handled by h2o since backend uses HTTP/1.1 without any knowledge of stream concurrency.

**Comment by kazuho** [comment]
Thank you for reporting the issue.  Confirmed that the problem exists.

It happens when H2O receives the end of the response from upstream server before h2spec succeeds in exceeding the maximum number of concurrent streams.

Contrary to the HTTP/2 spec, the pull-type handler of H2O (which is used by its reverse proxy implementation) decrements the number of open streams when it receives the end of the response from upstream servers.  This is the cause of the issue.

**Comment by kazuho** [comment]
fixed in #352.
